src/Npgsql/NpgsqlBatchCommandCollection.cs

```diff
@@ -97,8 +99,16 @@ protected override void SetBatchCommand(int index, DbBatchCommand batchCommand)
         => _list[index] = Cast(batchCommand);
 
     static NpgsqlBatchCommand Cast(DbBatchCommand? value)
-        => value is NpgsqlBatchCommand c
-            ? c
-            : throw new InvalidCastException(
-                $"The value \"{value}\" is not of type \"{nameof(NpgsqlBatchCommand)}\" and cannot be used in this batch command collection.");
+    {
+        var castedValue = value as NpgsqlBatchCommand;
+        if (castedValue is null)
+            ThrowInvalidCastException(value);
+
+        return castedValue;
+    }
+
+    [DoesNotReturn]
+    static void ThrowInvalidCastException(DbBatchCommand? value) =>
```

Comment:
Move it to `ThrowHelper` static class.

Reply:
Note that the exception message is very specific to NpgsqlBatchCommandCollection. We could pass it as an argument to ThrowInvalidCastException, as long as it doesn't produce worse code (which I don't think it should).

Reply:
> Note that the exception message is very specific to NpgsqlBatchCommandCollection. We could pass it as an argument to ThrowInvalidCastException, as long as it doesn't produce worse code (which I don't think it should).

It does become slightly worse in case of passing `NpgsqlParameter` to `NpgsqlParameterCollection.Add(object)`. JIT does inline it, and as such removes the throw altogether, but some of the remnants of string allocation remain :(
Specifically, I see:

```asm
sub       rsp,48
vxorps    xmm4,xmm4,xmm4
vmovdqa   xmmword ptr [rsp+20],xmm4
vmovdqa   xmmword ptr [rsp+30],xmm4
xor       eax,eax
mov       [rsp+40],rax
```